python_modules/libraries/dagster-powerbi/dagster_powerbi/translator.py

```diff
@@ -73,14 +91,70 @@ def get_asset_spec(self, data: PowerBIContentData) -> AssetSpec:
         else:
             check.assert_never(data.content_type)
 
-    def get_dashboard_asset_key(self, data: PowerBIContentData) -> AssetKey: ...
-    def get_dashboard_spec(self, data: PowerBIContentData) -> AssetSpec: ...
-
-    def get_report_asset_key(self, data: PowerBIContentData) -> AssetKey: ...
-    def get_report_spec(self, data: PowerBIContentData) -> AssetSpec: ...
-
-    def get_semantic_model_asset_key(self, data: PowerBIContentData) -> AssetKey: ...
-    def get_semantic_model_spec(self, data: PowerBIContentData) -> AssetSpec: ...
-
-    def get_data_source_asset_key(self, data: PowerBIContentData) -> AssetKey: ...
-    def get_data_source_spec(self, data: PowerBIContentData) -> AssetSpec: ...
+    def get_dashboard_asset_key(self, data: PowerBIContentData) -> AssetKey:
+        return AssetKey(
+            ["dashboard", _clean_asset_name(_remove_file_ext(data.properties["displayName"]))]
+        )
+
+    def get_dashboard_spec(self, data: PowerBIContentData) -> AssetSpec:
+        tile_report_ids = [
+            tile["reportId"] for tile in data.properties["tiles"] if "reportId" in tile
+        ]
+        report_keys = [
+            self.get_report_asset_key(self.workspace_data.reports_by_id[report_id])
+            for report_id in tile_report_ids
+        ]
+
+        return AssetSpec(
+            key=self.get_dashboard_asset_key(data),
+            tags={"dagster/storage_kind": "powerbi"},
+            deps=report_keys,
+        )
+
+    def get_report_asset_key(self, data: PowerBIContentData) -> AssetKey:
+        return AssetKey(["report", _clean_asset_name(data.properties["name"])])
+
+    def get_report_spec(self, data: PowerBIContentData) -> AssetSpec:
+        dataset_id = data.properties["datasetId"]
+        dataset_data = self.workspace_data.semantic_models_by_id.get(dataset_id)
+        dataset_key = self.get_semantic_model_asset_key(dataset_data) if dataset_data else None
+
+        return AssetSpec(
+            key=self.get_report_asset_key(data),
+            deps=[dataset_key] if dataset_key else None,
+            tags={"dagster/storage_kind": "powerbi"},
+        )
+
+    def get_semantic_model_asset_key(self, data: PowerBIContentData) -> AssetKey:
+        return AssetKey(["semantic_model", _clean_asset_name(data.properties["name"])])
```

Comment:
It's not super typical to include the asset type in the asset key. Is "semantic_model" part of the "address" in power BI?

Reply:
Insomuch as it's part of the URL, yes, the object types are prefixed with their type. Because many objects share names (e.g. semantic models and their raw data and semantic models and reports built on them) our options are:

1) Come up with a renaming strategy we apply to all objects by default, e.g. `report_customers`, `semantic_model_customers` etc
2) Use key prefixes to distinguish object types

Since there's no other real notion of namespacing here I am leaning on (2) rather than (1) because (1) involves us making a more deliberate departure from the raw data, but can change later.